### PR TITLE
Add mask support for some sparse common ops

### PR DESCRIPTION
**Pull request by FlorentinD**
Including:
- eWiseAdd
- eWiseMult
- mxm (sparse x sparse) + example
- reduceColumnWise

I plan to add benchmarks in a seperate PR (or after a first review)

### CODE REVIEW

**Comment by FlorentinD** [comment]
The last two commits include an optimization in resizing the result matrix. If you want I can also put them in a seperate PR.

File: examples/src/org/ejml/example/ExampleMaskedTriangleCount.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.ejml.dense.row.CommonOps_DDRM;
+import org.ejml.masks.DMaskFactory;
+import org.ejml.ops.DMonoids;
+import org.ejml.ops.DSemiRing;
+import org.ejml.ops.DSemiRings;
+import org.ejml.sparse.csc.CommonOpsWithSemiRing_DSCC;
+import org.ejml.sparse.csc.CommonOps_DSCC;
+
+/**
+ * Example using masked matrix multiplication to count the triangles in a graph.
```
Comment (lessthanoptimal) [inline examples/src/org/ejml/example/ExampleMaskedTriangleCount.java:31]: Is it possible to add a sentence with a more every day example of what triangles in a graph might be and how it would be useful? Maybe mention a problem this is important to?

File: main/ejml-dsparse/src/org/ejml/sparse/csc/CommonOps_DSCC.java [hunk 2]
```diff
@@ -1950,8 +1953,10 @@ public static DMatrixRMaj reduceColumnWise( DMatrixSparseCSC input, double initV
             int end = input.col_idx[col + 1];
 
             double acc = initValue;
-            for (int i = start; i < end; i++) {
-                acc = func.apply(acc, input.nz_values[i]);
+            if (mask == null || mask.isSet(col)) {
```
Comment (lessthanoptimal) [inline main/ejml-dsparse/src/org/ejml/sparse/csc/CommonOps_DSCC.java:1956]: Did needing to check the mask for null cause any performance regression?

File: main/ejml-dsparse/src/org/ejml/sparse/csc/misc/ImplCommonOpsWithSemiRing_DSCC.java [hunk 3]
```diff
@@ -146,9 +158,10 @@ public static void elementMult( DMatrixSparseCSC A, DMatrixSparseCSC B, DMatrixS
 
             // compute the maximum number of elements that there can be in this row
             int maxInRow = Math.min(idxA1 - idxA0, idxB1 - idxB0);
+            int expectedResultSize = C.nz_length + maxInRow;
 
             // make sure there are enough non-zero elements in C
-            if (C.nz_length + maxInRow > C.nz_values.length)
+            if (expectedResultSize > C.nz_values.length && expectedResultSize < maxMaskEntries)
```
Comment (lessthanoptimal) [inline main/ejml-dsparse/src/org/ejml/sparse/csc/misc/ImplCommonOpsWithSemiRing_DSCC.java]: I've not gone through the logic here carefully, but this jumps out as a potential issue. Are there unit tests which would trigger the logic on the right and have you checked edge conditions?

**Review by lessthanoptimal — COMMENTED** [review]
A couple of comments and nothing major.  I'm going to take another pass and try to go over some of the logic more carefully.

File: examples/src/org/ejml/example/ExampleMaskedTriangleCount.java (hunk 1 above)
Comment (FlorentinD) [reply examples/src/org/ejml/example/ExampleMaskedTriangleCount.java:31]: Sure

File: main/ejml-dsparse/src/org/ejml/sparse/csc/CommonOps_DSCC.java (hunk 2 above)
Comment (FlorentinD) [reply main/ejml-dsparse/src/org/ejml/sparse/csc/CommonOps_DSCC.java:1956]: Good point, ideally the JIT compiler can optimize the null check out.
But I can do a benchmark (next Monday after eastern).

File: main/ejml-dsparse/src/org/ejml/sparse/csc/misc/ImplCommonOpsWithSemiRing_DSCC.java (hunk 3 above)
Comment (FlorentinD) [reply main/ejml-dsparse/src/org/ejml/sparse/csc/misc/ImplCommonOpsWithSemiRing_DSCC.java]: I have the dedicated test `TestImplCommonOpsWithSemiRing_DSCC#useMaskForResultExpansionInMult` for this condition.
The idea is, the mask provides a __pessimistic__ estimation of its set entries. Thereby the number of set entries corresponds to the largest possible result size. The estimation will never be smaller than the actual number of set entries. 
So with this check we avoid growing larger than necessary.

Explaining the check, I also see this condition is hard to read. I will simplify the check :+1:

**Comment by lessthanoptimal** [comment]
FYI I'm planning on doing the second pass early this coming week. Thanks for the follow up comments!

File: main/ejml-dsparse/src/org/ejml/sparse/csc/CommonOps_DSCC.java (hunk 2 above)
Comment (FlorentinD) [reply main/ejml-dsparse/src/org/ejml/sparse/csc/CommonOps_DSCC.java:1956]: My benchmark showed no regression for this null check. Also worth to point out, the check is only done once per column. It should be different for row-wise reduction or apply.

File: main/ejml-dsparse/test/org/ejml/sparse/csc/misc/TestImplCommonOpsWithSemiRing_DSCC.java [hunk 4]
```diff
@@ -46,17 +50,68 @@ public void add( DSemiRing semiRing, double[] expected) {
...
+        var a = RandomMatrices_DSCC.rectangle(10, 10, 30, random);
+        var b = RandomMatrices_DSCC.rectangle(10, 10, 30, random);
+        var mask = DMaskFactory.builder(RandomMatrices_DSCC.rectangle(10, 10, 30, random), true).build();
+
+        var unmasked = new DMatrixSparseCSC(10, 10, 0);
+        var masked = new DMatrixSparseCSC(10, 10, 0);
+
+        ImplCommonOpsWithSemiRing_DSCC.add(1, a, 1, b, unmasked, DSemiRings.PLUS_TIMES, null, null, null);
+        ImplCommonOpsWithSemiRing_DSCC.add(1, a, 1, b, masked, DSemiRings.PLUS_TIMES, mask, null, null);
+
```
Comment (lessthanoptimal) [inline main/ejml-dsparse/test/org/ejml/sparse/csc/misc/TestImplCommonOpsWithSemiRing_DSCC.java:73]: extra space

File: main/ejml-dsparse/test/org/ejml/sparse/csc/misc/TestImplCommonOpsWithSemiRing_DSCC.java [hunk 5]
```diff
@@ -71,22 +126,39 @@ public void testelementWiseMult( DSemiRing semiRing, double[] expected) {
...
+        var a = RandomMatrices_DSCC.rectangle(10, 10, 30, random);
+        var b = RandomMatrices_DSCC.rectangle(10, 10, 30, random);
+        var mask = DMaskFactory.builder(RandomMatrices_DSCC.rectangle(10, 10, 30, random), true).build();
+
+        var unmasked = new DMatrixSparseCSC(10, 10, 0);
+        var masked = new DMatrixSparseCSC(10, 10, 0);
+
+        ImplCommonOpsWithSemiRing_DSCC.elementMult(a, b, unmasked, DSemiRings.PLUS_TIMES, null, null, null);
+        ImplCommonOpsWithSemiRing_DSCC.elementMult(a, b, masked, DSemiRings.PLUS_TIMES, mask, null, null);
+
```
Comment (lessthanoptimal) [inline main/ejml-dsparse/test/org/ejml/sparse/csc/misc/TestImplCommonOpsWithSemiRing_DSCC.java:148]: extra space/line here too

File: main/ejml-dsparse/test/org/ejml/sparse/csc/mult/TestMatrixMatrixMultWithSemiRing_DSCC.java [hunk 6]
```diff
@@ -89,12 +92,28 @@ void elementMult(String desc, DMatrixSparseCSC matrix, DMatrixSparseCSC otherMat
     void add(String desc, DMatrixSparseCSC matrix, DMatrixSparseCSC otherMatrix) {
         DSemiRing semiRing = DSemiRings.PLUS_TIMES;
 
-        DMatrixSparseCSC found = CommonOpsWithSemiRing_DSCC.add(1, matrix, 1, otherMatrix, null, semiRing, null, null);
+        DMatrixSparseCSC found = CommonOpsWithSemiRing_DSCC.add(1, matrix, 1, otherMatrix, null, semiRing, null,null, null);
```
Comment (lessthanoptimal) [inline main/ejml-dsparse/test/org/ejml/sparse/csc/mult/TestMatrixMatrixMultWithSemiRing_DSCC.java]: don't forget to run the code formatter on these files. not sure when there will be a lint check for this

**Review by lessthanoptimal — APPROVED** [review]
some minor nits. just clean up the formatting a little bit. looks good and thanks for the PR! Sorry for taking so long to go over it.

File: main/ejml-dsparse/src/org/ejml/sparse/csc/misc/ImplCommonOpsWithSemiRing_DSCC.java (hunk 3 above)
Comment (lessthanoptimal) [reply main/ejml-dsparse/src/org/ejml/sparse/csc/misc/ImplCommonOpsWithSemiRing_DSCC.java]: thanks! easier to understand now.

Comment (lessthanoptimal) [reply main/ejml-dsparse/src/org/ejml/sparse/csc/misc/ImplCommonOpsWithSemiRing_DSCC.java]: If this growth logic occurs in several locations we should probably create a static function that does it and create a new PR for it.

Comment (FlorentinD) [reply main/ejml-dsparse/src/org/ejml/sparse/csc/misc/ImplCommonOpsWithSemiRing_DSCC.java]: Yes I think this could be beneficial when we add more sparse operations with mask support
